#56 integrated SMCloudStore with Azure provider; cleaned Item form in frontend

Anjali0407-git opened the pull request:
**Fixes #56**

### Changes Made:

- Integrated the SMCloudStore library into the Node.js server to establish connections with the Azure Storage provider.

### Reasons for Changes:

- To leverage the scalability and flexibility of cloud storage solutions, integrating with a cloud provider like Azure allows efficient storage and retrieval of image files. This enhances our application's ability to handle media files.

### Implementation Details:

- Set up a new Azure storage account and configured it using the SMCloudStore API for seamless integration.
- Updated the Node.js server to handle image uploads: images are now received via POST requests and uploaded directly to Azure Storage.
- Modified the database schema to store image URLs in the `container-name/file-name` format, facilitating easier and more efficient retrieval.
- Refactored the frontend to streamline the process of sending images. This involved adjusting the form data handling and ensuring compatibility with the new backend setup.


### Screenshots

- **Files uploaded to Azure storage**
  
  ![image](https://github.com/user-attachments/assets/d642071c-2827-40c4-9b81-23de87fea9c3)

- **Postgres storing list of filenames from azure storage**

  ![image](https://github.com/user-attachments/assets/963b2f39-4b5d-49e7-9fc3-5b47827295bc)

kungfuchicken commented on server/src/configs/SMCloudStoreConfig.ts:
starting with Azure makes sense based on what we discussed. do you have an issue created to add an option for all the supported platforms? it seems like it would be a straightforward update to this file and the donated item service.

kungfuchicken replied:
i guess package.json, too 😆

Anjali0407-git replied:
writing a new connection for other cloud storage is easy but I have to create accounts for other cloud providers to get the connection keys, storage account etc which is time consuming. I will create an issue for that. What is the purpose of the multiple cloud providers? Is it for fail safe or to be able to switch to other providers in future?

kungfuchicken replied:
to switch to other providers in the future, especially if the solution is deployed by other users. solving for BWorks is our first use case, but in talking w/ Patrick my understanding is that his vision has always been something that anyone who does similar things could re-use the tool.

Anjali0407-git replied:
It makes sense. I will work on that.

kungfuchicken commented on client-app/src/Components/NewItemForm.tsx:
i don't see any validation that you're only accepting images, indication what sort of images types might be accepted, or limits on file size. these aren't urgent concerns, but you'll probably want them at least in the client, and probably in the server, at some point. maybe just drop a new issue.

Anjali0407-git replied:
It makes sense, Copied!!

kungfuchicken reviewed (commented):
This looks great! Please either address the three comments with either a bit more code, or new Issues.

kungfuchicken commented:
Jest is failing, too: https://github.com/oss-slu/material-donor-mutual-assist/actions/runs/11522648230/job/32078965778?pr=58#step:5:68

Anjali0407-git commented on server/src/routes/donatedItemRoutes.ts:
```diff
@@ -27,16 +32,24 @@ router.post('/', donatedItemValidator, async (req: Request, res: Response) => {
+        // upload images to Azure and get their filenames
+        const imageUrls = await Promise.all(imageFiles.map((file, index) => {
+            const formattedDate = new Date().toISOString();
+            return uploadToAzure(file, `item-${formattedDate}-${newItem.id}.jpg`);
```
Ah it makes sense. I blindly used .jpg but it could be .png image also. I will try to fix it or create a good first issue.

Anjali0407-git replied:
Fixed this one by creating a function to find the file extension based on incoming file MIME type.

Anjali0407-git commented:
I have addressed one of the review comments and updated unit tests and now CI pipeline has passed.
And I created the following issues to address other two review comments.
[Issue1](https://github.com/orgs/oss-slu/projects/41/views/1?pane=issue&itemId=85140626)
[Issue2](https://github.com/orgs/oss-slu/projects/41/views/1?pane=issue&itemId=85139047)
[Issue3](https://github.com/orgs/oss-slu/projects/41/views/1?pane=issue&itemId=85139787)

kungfuchicken reviewed (approved):
lgtm; Thank you!
